### ISSUE TITLE
build failed

### ISSUE THREAD

**Issue by eleven-pixel** [supranational/supra_seal#1]
/filecoin/supra_seal$ ./build.sh 
+ SECTOR_SIZE=-DSECTOR_SIZE_32GiB
+ '[' '' == 512MiB ']'
+ CUDA=/usr/local/cuda
+ NVCC=/usr/local/cuda/bin/nvcc
+ SPDK=deps/spdk-v22.09
+ INCLUDE='-Ideps/spdk-v22.09/include -Ideps/spdk-v22.09/isa-l/.. -Ideps/spdk-v22.09/dpdk/build/include'
+ CFLAGS='-DSECTOR_SIZE_32GiB -g -Ideps/spdk-v22.09/include -Ideps/spdk-v22.09/isa-l/.. -Ideps/spdk-v22.09/dpdk/build/include'
+ CPPFLAGS='-DSECTOR_SIZE_32GiB -g -Ideps/spdk-v22.09/include -Ideps/spdk-v22.09/isa-l/.. -Ideps/spdk-v22.09/dpdk/build/include           -fno-omit-frame-pointer -g -O2 -Wall -Wextra -Wno-unused-parameter           -Wno-missing-field-initializers -fno-strict-aliasing           -march=native -Wformat -Wformat-security           -D_GNU_SOURCE -fPIC -fstack-protector           -fno-common -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2           -DSPDK_GIT_COMMIT=4be6d3043 -pthread'
+ LDFLAGS='-fno-omit-frame-pointer -Wl,-z,relro,-z,now -Wl,-z,noexecstack -fuse-ld=bfd         -Ldeps/spdk-v22.09/build/lib          -Wl,--whole-archive -Wl,--no-as-needed          -lspdk_bdev_malloc          -lspdk_bdev_null          -lspdk_bdev_nvme          -lspdk_bdev_passthru          -lspdk_bdev_lvol          -lspdk_bdev_raid          -lspdk_bdev_error          -lspdk_bdev_gpt          -lspdk_bdev_split          -lspdk_bdev_delay          -lspdk_bdev_zone_block          -lspdk_blobfs_bdev          -lspdk_blobfs          -lspdk_blob_bdev          -lspdk_lvol          -lspdk_blob          -lspdk_nvme          -lspdk_bdev_ftl          -lspdk_ftl          -lspdk_bdev_aio          -lspdk_bdev_virtio          -lspdk_virtio          -lspdk_vfio_user          -lspdk_accel_ioat          -lspdk_ioat          -lspdk_scheduler_dynamic          -lspdk_env_dpdk          -lspdk_scheduler_dpdk_governor          -lspdk_scheduler_gscheduler          -lspdk_sock_posix          -lspdk_event          -lspdk_event_bdev          -lspdk_bdev          -lspdk_notify          -lspdk_dma          -lspdk_event_accel          -lspdk_accel          -lspdk_event_vmd          -lspdk_vmd          -lspdk_event_sock          -lspdk_init          -lspdk_thread          -lspdk_trace          -lspdk_sock          -lspdk_rpc          -lspdk_jsonrpc          -lspdk_json          -lspdk_util          -lspdk_log          -Wl,--no-whole-archive deps/spdk-v22.09/build/lib/libspdk_env_dpdk.a          -Wl,--whole-archive deps/spdk-v22.09/dpdk/build/lib/librte_bus_pci.a          deps/spdk-v22.09/dpdk/build/lib/librte_cryptodev.a          deps/spdk-v22.09/dpdk/build/lib/librte_dmadev.a          deps/spdk-v22.09/dpdk/build/lib/librte_eal.a          deps/spdk-v22.09/dpdk/build/lib/librte_ethdev.a          deps/spdk-v22.09/dpdk/build/lib/librte_hash.a          deps/spdk-v22.09/dpdk/build/lib/librte_kvargs.a          deps/spdk-v22.09/dpdk/build/lib/librte_mbuf.a          deps/spdk-v22.09/dpdk/build/lib/librte_mempool.a          deps/spdk-v22.09/dpdk/build/lib/librte_mempool_ring.a          deps/spdk-v22.09/dpdk/build/lib/librte_net.a          deps/spdk-v22.09/dpdk/build/lib/librte_pci.a          deps/spdk-v22.09/dpdk/build/lib/librte_power.a          deps/spdk-v22.09/dpdk/build/lib/librte_rcu.a          deps/spdk-v22.09/dpdk/build/lib/librte_ring.a          deps/spdk-v22.09/dpdk/build/lib/librte_telemetry.a          deps/spdk-v22.09/dpdk/build/lib/librte_vhost.a          -Wl,--no-whole-archive          -lnuma -ldl          -Ldeps/spdk-v22.09/isa-l/.libs -lisal          -pthread -lrt -luuid -lssl -lcrypto -lm -laio'
+ '[' '!' -d /var/tmp/supra_seal ']'
+ rm -fr obj
+ mkdir -p obj
+ mkdir -p deps
+ '[' '!' -d deps/spdk-v22.09 ']'
+ '[' '!' -d deps/sppark ']'
+ '[' '!' -d deps/blst ']'
+ gcc -c src/sha/sha_ext_mbx2.S -o obj/sha_ext_mbx2.o
+ xxd -i src/poseidon/constants/constants_2
+ xxd -i src/poseidon/constants/constants_4
+ xxd -i src/poseidon/constants/constants_8
+ xxd -i src/poseidon/constants/constants_11
+ xxd -i src/poseidon/constants/constants_16
+ xxd -i src/poseidon/constants/constants_24
+ xxd -i src/poseidon/constants/constants_36
+ c++ -DSECTOR_SIZE_32GiB -g -Ideps/spdk-v22.09/include -Ideps/spdk-v22.09/isa-l/.. -Ideps/spdk-v22.09/dpdk/build/include -fno-omit-frame-pointer -g -O2 -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -fno-strict-aliasing -march=native -Wformat -Wformat-security -D_GNU_SOURCE -fPIC -fstack-protector -fno-common -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -DSPDK_GIT_COMMIT=4be6d3043 -pthread -o obj/pc1.o -c src/pc1/pc1.cpp
+ c++ -DSECTOR_SIZE_32GiB -g -Ideps/spdk-v22.09/include -Ideps/spdk-v22.09/isa-l/.. -Ideps/spdk-v22.09/dpdk/build/include -fno-omit-frame-pointer -g -O2 -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -fno-strict-aliasing -march=native -Wformat -Wformat-security -D_GNU_SOURCE -fPIC -fstack-protector -fno-common -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -DSPDK_GIT_COMMIT=4be6d3043 -pthread -o obj/column_reader.o -c src/pc2/column_reader.cpp
+ /usr/local/cuda/bin/nvcc -DSECTOR_SIZE_32GiB -g -Ideps/spdk-v22.09/include -Ideps/spdk-v22.09/isa-l/.. -Ideps/spdk-v22.09/dpdk/build/include -std=c++17 -DNO_SPDK -D__ADX__ -Xcompiler -Wno-subobject-linkage -Xcompiler -O2 -Ideps/sppark -Ideps/blst/src -arch=sm_75 -dc src/pc2/cuda/pc2.cu -o obj/pc2.o
./build.sh: line 152: /usr/local/cuda/bin/nvcc: No such file or directory
+ /usr/local/cuda/bin/nvcc -DSECTOR_SIZE_32GiB -g -Ideps/spdk-v22.09/include -Ideps/spdk-v22.09/isa-l/.. -Ideps/spdk-v22.09/dpdk/build/include -std=c++17 -DNO_SPDK -D__ADX__ -Xcompiler -Wno-subobject-linkage -Xcompiler -O2 -Ideps/sppark -Ideps/blst/src -arch=sm_75 -dlink src/pc2/cuda/pc2.cu -o obj/pc2_link.o
+ c++ -c src/sealing/sector_parameters.cpp -o obj/sector_parameters.o
./build.sh: line 154: /usr/local/cuda/bin/nvcc: No such file or directory
+ c++ -DSECTOR_SIZE_32GiB -g -Ideps/spdk-v22.09/include -Ideps/spdk-v22.09/isa-l/.. -Ideps/spdk-v22.09/dpdk/build/include -fno-omit-frame-pointer -g -O2 -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -fno-strict-aliasing -march=native -Wformat -Wformat-security -D_GNU_SOURCE -fPIC -fstack-protector -fno-common -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -DSPDK_GIT_COMMIT=4be6d3043 -pthread -c src/c1/c1.cpp -o obj/c1.o -Wno-subobject-linkage -Ideps/sppark -Ideps/blst/src
+ wait
+ c++ -DSECTOR_SIZE_32GiB -g -Ideps/spdk-v22.09/include -Ideps/spdk-v22.09/isa-l/.. -Ideps/spdk-v22.09/dpdk/build/include -fno-omit-frame-pointer -g -O2 -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -fno-strict-aliasing -march=native -Wformat -Wformat-security -D_GNU_SOURCE -fPIC -fstack-protector -fno-common -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -DSPDK_GIT_COMMIT=4be6d3043 -pthread -Ideps/spdk-v22.09/include -Ideps/spdk-v22.09/isa-l/.. -Ideps/spdk-v22.09/dpdk/build/include -c src/sealing/supra_seal.cpp -o obj/supra_seal.o -Wno-subobject-linkage
In file included from src/pc1/pc1.hpp:6:0,
                 from src/pc1/pc1.cpp:31:
src/pc1/../sealing/topology_t.hpp: In constructor ‘topology_t::topology_t(const char*)’:
src/pc1/../sealing/topology_t.hpp:145:35: error: missing template arguments before ‘(’ token
         sector_configs.insert(pair((size_t)sector_config.sectors, sector_config));
                                   ^
In file included from src/sealing/../pc1/pc1.hpp:6:0,
                 from src/sealing/supra_seal.cpp:31:
src/sealing/../pc1/../sealing/topology_t.hpp: In constructor ‘topology_t::topology_t(const char*)’:
src/sealing/../pc1/../sealing/topology_t.hpp:145:35: error: missing template arguments before ‘(’ token
         sector_configs.insert(pair((size_t)sector_config.sectors, sector_config));
                                   ^
src/pc1/pc1.cpp: In lambda function:
src/pc1/pc1.cpp:143:21: error: missing template arguments before ‘coordinator’
       coordinator_t coordinator(terminator, system,
                     ^~~~~~~~~~~
src/pc1/pc1.cpp:148:40: error: ‘coordinator’ was not declared in this scope
       system.coordinators[coord_id] = &coordinator;
                                        ^~~~~~~~~~~
src/pc1/pc1.cpp:148:40: note: suggested alternative: ‘coordinator_t’
       system.coordinators[coord_id] = &coordinator;
                                        ^~~~~~~~~~~
                                        coordinator_t
In file included from src/pc1/pc1.cpp:50:0:
src/pc1/system_buffers_t.hpp: In instantiation of ‘int system_buffers_t<C>::init(size_t) [with C = sealing_config_t<128>; size_t = long unsigned int]’:
src/pc1/pc1.cpp:83:3:   required from ‘int do_pc1(nvme_controllers_t*, topology_t&, uint64_t, const uint32_t*, const char*) [with C = sealing_config_t<128>; uint64_t = long unsigned int; uint32_t = unsigned int]’
src/pc1/pc1.cpp:183:70:   required from here
src/pc1/system_buffers_t.hpp:176:29: error: use of deleted function ‘std::atomic<_Tp>::atomic() [with _Tp = node_id_t]’
       coordinator_node[i] = new atomic<node_id_t>();
                             ^~~~~~~~~~~~~~~~~~~~~~~
In file included from src/pc1/../sealing/../util/thread_pool_t.hpp:13:0,
                 from src/pc1/../sealing/topology_t.hpp:10,
                 from src/pc1/pc1.hpp:6,
                 from src/pc1/pc1.cpp:31:
/usr/include/c++/7/atomic:194:7: note: ‘std::atomic<_Tp>::atomic() noexcept [with _Tp = node_id_t]’ is implicitly deleted because its exception-specification does not match the implicit exception-specification ‘’
       atomic() noexcept = default;
       ^~~~~~
In file included from src/pc1/pc1.cpp:50:0:
src/pc1/system_buffers_t.hpp: In instantiation of ‘int system_buffers_t<C>::init(size_t) [with C = sealing_config_t<64>; size_t = long unsigned int]’:
src/pc1/pc1.cpp:83:3:   required from ‘int do_pc1(nvme_controllers_t*, topology_t&, uint64_t, const uint32_t*, const char*) [with C = sealing_config_t<64>; uint64_t = long unsigned int; uint32_t = unsigned int]’
src/pc1/pc1.cpp:188:69:   required from here
src/pc1/system_buffers_t.hpp:176:29: error: use of deleted function ‘std::atomic<_Tp>::atomic() [with _Tp = node_id_t]’
       coordinator_node[i] = new atomic<node_id_t>();
                             ^~~~~~~~~~~~~~~~~~~~~~~
src/pc1/system_buffers_t.hpp: In instantiation of ‘int system_buffers_t<C>::init(size_t) [with C = sealing_config_t<32>; size_t = long unsigned int]’:
src/pc1/pc1.cpp:83:3:   required from ‘int do_pc1(nvme_controllers_t*, topology_t&, uint64_t, const uint32_t*, const char*) [with C = sealing_config_t<32>; uint64_t = long unsigned int; uint32_t = unsigned int]’
src/pc1/pc1.cpp:193:69:   required from here
src/pc1/system_buffers_t.hpp:176:29: error: use of deleted function ‘std::atomic<_Tp>::atomic() [with _Tp = node_id_t]’
src/pc1/system_buffers_t.hpp: In instantiation of ‘int system_buffers_t<C>::init(size_t) [with C = sealing_config_t<16>; size_t = long unsigned int]’:
src/pc1/pc1.cpp:83:3:   required from ‘int do_pc1(nvme_controllers_t*, topology_t&, uint64_t, const uint32_t*, const char*) [with C = sealing_config_t<16>; uint64_t = long unsigned int; uint32_t = unsigned int]’
src/pc1/pc1.cpp:198:69:   required from here
src/pc1/system_buffers_t.hpp:176:29: error: use of deleted function ‘std::atomic<_Tp>::atomic() [with _Tp = node_id_t]’
src/pc1/system_buffers_t.hpp: In instantiation of ‘int system_buffers_t<C>::init(size_t) [with C = sealing_config_t<8>; size_t = long unsigned int]’:
src/pc1/pc1.cpp:83:3:   required from ‘int do_pc1(nvme_controllers_t*, topology_t&, uint64_t, const uint32_t*, const char*) [with C = sealing_config_t<8>; uint64_t = long unsigned int; uint32_t = unsigned int]’
src/pc1/pc1.cpp:203:68:   required from here
src/pc1/system_buffers_t.hpp:176:29: error: use of deleted function ‘std::atomic<_Tp>::atomic() [with _Tp = node_id_t]’
src/pc1/system_buffers_t.hpp: In instantiation of ‘int system_buffers_t<C>::init(size_t) [with C = sealing_config_t<4>; size_t = long unsigned int]’:
src/pc1/pc1.cpp:83:3:   required from ‘int do_pc1(nvme_controllers_t*, topology_t&, uint64_t, const uint32_t*, const char*) [with C = sealing_config_t<4>; uint64_t = long unsigned int; uint32_t = unsigned int]’
src/pc1/pc1.cpp:208:68:   required from here
src/pc1/system_buffers_t.hpp:176:29: error: use of deleted function ‘std::atomic<_Tp>::atomic() [with _Tp = node_id_t]’
src/pc1/system_buffers_t.hpp: In instantiation of ‘int system_buffers_t<C>::init(size_t) [with C = sealing_config_t<2>; size_t = long unsigned int]’:
src/pc1/pc1.cpp:83:3:   required from ‘int do_pc1(nvme_controllers_t*, topology_t&, uint64_t, const uint32_t*, const char*) [with C = sealing_config_t<2>; uint64_t = long unsigned int; uint32_t = unsigned int]’
src/pc1/pc1.cpp:213:68:   required from here
src/pc1/system_buffers_t.hpp:176:29: error: use of deleted function ‘std::atomic<_Tp>::atomic() [with _Tp = node_id_t]’
In file included from src/pc1/../nvme/nvme.hpp:20:0,
                 from src/pc1/pc1.cpp:36:
src/pc1/../nvme/ring_t.hpp: In instantiation of ‘ring_buffer_valid_t ring_buffer_t<T, _VALID_THRESHOLD, SIZE>::get_valid(size_t) [with T = batch_t<page_t<sealing_config_t<128> >, 1>; unsigned int _VALID_THRESHOLD = 1; int SIZE = 1048576; ring_buffer_valid_t = std::atomic<long unsigned int>; size_t = long unsigned int]’:
src/pc1/orchestrator_t.hpp:687:61:   required from ‘int orchestrator_t<C>::process(bool) [with C = sealing_config_t<128>]’
src/pc1/pc1.cpp:161:3:   required from ‘int do_pc1(nvme_controllers_t*, topology_t&, uint64_t, const uint32_t*, const char*) [with C = sealing_config_t<128>; uint64_t = long unsigned int; uint32_t = unsigned int]’
src/pc1/pc1.cpp:183:70:   required from here
src/pc1/../nvme/ring_t.hpp:397:48: error: use of deleted function ‘std::atomic<long unsigned int>::atomic(const std::atomic<long unsigned int>&)’
     return valid[idx].load(DEFAULT_MEMORY_ORDER);
                                                ^
In file included from src/pc1/../sealing/../util/thread_pool_t.hpp:13:0,
                 from src/pc1/../sealing/topology_t.hpp:10,
                 from src/pc1/pc1.hpp:6,
                 from src/pc1/pc1.cpp:31:
/usr/include/c++/7/atomic:737:7: note: declared here
       atomic(const atomic&) = delete;
       ^~~~~~
/usr/include/c++/7/atomic:741:17: note:   after user-defined conversion: constexpr std::atomic<long unsigned int>::atomic(std::atomic<long unsigned int>::__integral_type)
       constexpr atomic(__integral_type __i) noexcept : __base_type(__i) { }
                 ^~~~~~
In file included from src/pc1/../nvme/nvme.hpp:20:0,
                 from src/pc1/pc1.cpp:36:
src/pc1/../nvme/ring_t.hpp: In instantiation of ‘ring_buffer_valid_t ring_buffer_t<T, _VALID_THRESHOLD, SIZE>::get_valid(size_t) [with T = batch_t<page_t<sealing_config_t<64> >, 1>; unsigned int _VALID_THRESHOLD = 1; int SIZE = 1048576; ring_buffer_valid_t = std::atomic<long unsigned int>; size_t = long unsigned int]’:
src/pc1/orchestrator_t.hpp:687:61:   required from ‘int orchestrator_t<C>::process(bool) [with C = sealing_config_t<64>]’
src/pc1/pc1.cpp:161:3:   required from ‘int do_pc1(nvme_controllers_t*, topology_t&, uint64_t, const uint32_t*, const char*) [with C = sealing_config_t<64>; uint64_t = long unsigned int; uint32_t = unsigned int]’
src/pc1/pc1.cpp:188:69:   required from here
src/pc1/../nvme/ring_t.hpp:397:48: error: use of deleted function ‘std::atomic<long unsigned int>::atomic(const std::atomic<long unsigned int>&)’
     return valid[idx].load(DEFAULT_MEMORY_ORDER);
                                                ^
In file included from src/pc1/../sealing/../util/thread_pool_t.hpp:13:0,
                 from src/pc1/../sealing/topology_t.hpp:10,
                 from src/pc1/pc1.hpp:6,
                 from src/pc1/pc1.cpp:31:
/usr/include/c++/7/atomic:737:7: note: declared here
       atomic(const atomic&) = delete;
       ^~~~~~
/usr/include/c++/7/atomic:741:17: note:   after user-defined conversion: constexpr std::atomic<long unsigned int>::atomic(std::atomic<long unsigned int>::__integral_type)
       constexpr atomic(__integral_type __i) noexcept : __base_type(__i) { }
                 ^~~~~~
In file included from src/pc1/../nvme/nvme.hpp:20:0,
                 from src/pc1/pc1.cpp:36:
src/pc1/../nvme/ring_t.hpp: In instantiation of ‘ring_buffer_valid_t ring_buffer_t<T, _VALID_THRESHOLD, SIZE>::get_valid(size_t) [with T = batch_t<page_t<sealing_config_t<32> >, 1>; unsigned int _VALID_THRESHOLD = 1; int SIZE = 1048576; ring_buffer_valid_t = std::atomic<long unsigned int>; size_t = long unsigned int]’:
src/pc1/orchestrator_t.hpp:687:61:   required from ‘int orchestrator_t<C>::process(bool) [with C = sealing_config_t<32>]’
src/pc1/pc1.cpp:161:3:   required from ‘int do_pc1(nvme_controllers_t*, topology_t&, uint64_t, const uint32_t*, const char*) [with C = sealing_config_t<32>; uint64_t = long unsigned int; uint32_t = unsigned int]’
src/pc1/pc1.cpp:193:69:   required from here
src/pc1/../nvme/ring_t.hpp:397:48: error: use of deleted function ‘std::atomic<long unsigned int>::atomic(const std::atomic<long unsigned int>&)’
     return valid[idx].load(DEFAULT_MEMORY_ORDER);
                                                ^
In file included from src/pc1/../sealing/../util/thread_pool_t.hpp:13:0,
                 from src/pc1/../sealing/topology_t.hpp:10,
                 from src/pc1/pc1.hpp:6,
                 from src/pc1/pc1.cpp:31:
/usr/include/c++/7/atomic:737:7: note: declared here
       atomic(const atomic&) = delete;
       ^~~~~~
/usr/include/c++/7/atomic:741:17: note:   after user-defined conversion: constexpr std::atomic<long unsigned int>::atomic(std::atomic<long unsigned int>::__integral_type)
       constexpr atomic(__integral_type __i) noexcept : __base_type(__i) { }
                 ^~~~~~
In file included from src/pc1/../nvme/nvme.hpp:20:0,
                 from src/pc1/pc1.cpp:36:
src/pc1/../nvme/ring_t.hpp: In instantiation of ‘ring_buffer_valid_t ring_buffer_t<T, _VALID_THRESHOLD, SIZE>::get_valid(size_t) [with T = batch_t<page_t<sealing_config_t<16> >, 1>; unsigned int _VALID_THRESHOLD = 1; int SIZE = 1048576; ring_buffer_valid_t = std::atomic<long unsigned int>; size_t = long unsigned int]’:
src/pc1/orchestrator_t.hpp:687:61:   required from ‘int orchestrator_t<C>::process(bool) [with C = sealing_config_t<16>]’
src/pc1/pc1.cpp:161:3:   required from ‘int do_pc1(nvme_controllers_t*, topology_t&, uint64_t, const uint32_t*, const char*) [with C = sealing_config_t<16>; uint64_t = long unsigned int; uint32_t = unsigned int]’
src/pc1/pc1.cpp:198:69:   required from here
src/pc1/../nvme/ring_t.hpp:397:48: error: use of deleted function ‘std::atomic<long unsigned int>::atomic(const std::atomic<long unsigned int>&)’
     return valid[idx].load(DEFAULT_MEMORY_ORDER);
                                                ^
In file included from src/pc1/../sealing/../util/thread_pool_t.hpp:13:0,
                 from src/pc1/../sealing/topology_t.hpp:10,
                 from src/pc1/pc1.hpp:6,
                 from src/pc1/pc1.cpp:31:
/usr/include/c++/7/atomic:737:7: note: declared here
       atomic(const atomic&) = delete;
       ^~~~~~
/usr/include/c++/7/atomic:741:17: note:   after user-defined conversion: constexpr std::atomic<long unsigned int>::atomic(std::atomic<long unsigned int>::__integral_type)
       constexpr atomic(__integral_type __i) noexcept : __base_type(__i) { }
                 ^~~~~~
In file included from src/pc1/../nvme/nvme.hpp:20:0,
                 from src/pc1/pc1.cpp:36:
src/pc1/../nvme/ring_t.hpp: In instantiation of ‘ring_buffer_valid_t ring_buffer_t<T, _VALID_THRESHOLD, SIZE>::get_valid(size_t) [with T = batch_t<page_t<sealing_config_t<8> >, 1>; unsigned int _VALID_THRESHOLD = 1; int SIZE = 1048576; ring_buffer_valid_t = std::atomic<long unsigned int>; size_t = long unsigned int]’:
src/pc1/orchestrator_t.hpp:687:61:   required from ‘int orchestrator_t<C>::process(bool) [with C = sealing_config_t<8>]’
src/pc1/pc1.cpp:161:3:   required from ‘int do_pc1(nvme_controllers_t*, topology_t&, uint64_t, const uint32_t*, const char*) [with C = sealing_config_t<8>; uint64_t = long unsigned int; uint32_t = unsigned int]’
src/pc1/pc1.cpp:203:68:   required from here
src/pc1/../nvme/ring_t.hpp:397:48: error: use of deleted function ‘std::atomic<long unsigned int>::atomic(const std::atomic<long unsigned int>&)’
     return valid[idx].load(DEFAULT_MEMORY_ORDER);
                                                ^
In file included from src/pc1/../sealing/../util/thread_pool_t.hpp:13:0,
                 from src/pc1/../sealing/topology_t.hpp:10,
                 from src/pc1/pc1.hpp:6,
                 from src/pc1/pc1.cpp:31:
/usr/include/c++/7/atomic:737:7: note: declared here
       atomic(const atomic&) = delete;
       ^~~~~~
/usr/include/c++/7/atomic:741:17: note:   after user-defined conversion: constexpr std::atomic<long unsigned int>::atomic(std::atomic<long unsigned int>::__integral_type)
       constexpr atomic(__integral_type __i) noexcept : __base_type(__i) { }
                 ^~~~~~
In file included from src/pc1/../nvme/nvme.hpp:20:0,
                 from src/pc1/pc1.cpp:36:
src/pc1/../nvme/ring_t.hpp: In instantiation of ‘ring_buffer_valid_t ring_buffer_t<T, _VALID_THRESHOLD, SIZE>::get_valid(size_t) [with T = batch_t<page_t<sealing_config_t<4> >, 1>; unsigned int _VALID_THRESHOLD = 1; int SIZE = 1048576; ring_buffer_valid_t = std::atomic<long unsigned int>; size_t = long unsigned int]’:
src/pc1/orchestrator_t.hpp:687:61:   required from ‘int orchestrator_t<C>::process(bool) [with C = sealing_config_t<4>]’
src/pc1/pc1.cpp:161:3:   required from ‘int do_pc1(nvme_controllers_t*, topology_t&, uint64_t, const uint32_t*, const char*) [with C = sealing_config_t<4>; uint64_t = long unsigned int; uint32_t = unsigned int]’
src/pc1/pc1.cpp:208:68:   required from here
src/pc1/../nvme/ring_t.hpp:397:48: error: use of deleted function ‘std::atomic<long unsigned int>::atomic(const std::atomic<long unsigned int>&)’
     return valid[idx].load(DEFAULT_MEMORY_ORDER);
                                                ^
In file included from src/pc1/../sealing/../util/thread_pool_t.hpp:13:0,
                 from src/pc1/../sealing/topology_t.hpp:10,
                 from src/pc1/pc1.hpp:6,
                 from src/pc1/pc1.cpp:31:
/usr/include/c++/7/atomic:737:7: note: declared here
       atomic(const atomic&) = delete;
       ^~~~~~
/usr/include/c++/7/atomic:741:17: note:   after user-defined conversion: constexpr std::atomic<long unsigned int>::atomic(std::atomic<long unsigned int>::__integral_type)
       constexpr atomic(__integral_type __i) noexcept : __base_type(__i) { }
                 ^~~~~~
In file included from src/pc1/../nvme/nvme.hpp:20:0,
                 from src/pc1/pc1.cpp:36:
src/pc1/../nvme/ring_t.hpp: In instantiation of ‘ring_buffer_valid_t ring_buffer_t<T, _VALID_THRESHOLD, SIZE>::get_valid(size_t) [with T = batch_t<page_t<sealing_config_t<2> >, 1>; unsigned int _VALID_THRESHOLD = 1; int SIZE = 1048576; ring_buffer_valid_t = std::atomic<long unsigned int>; size_t = long unsigned int]’:
src/pc1/orchestrator_t.hpp:687:61:   required from ‘int orchestrator_t<C>::process(bool) [with C = sealing_config_t<2>]’
src/pc1/pc1.cpp:161:3:   required from ‘int do_pc1(nvme_controllers_t*, topology_t&, uint64_t, const uint32_t*, const char*) [with C = sealing_config_t<2>; uint64_t = long unsigned int; uint32_t = unsigned int]’
src/pc1/pc1.cpp:213:68:   required from here
src/pc1/../nvme/ring_t.hpp:397:48: error: use of deleted function ‘std::atomic<long unsigned int>::atomic(const std::atomic<long unsigned int>&)’
     return valid[idx].load(DEFAULT_MEMORY_ORDER);
                                                ^
In file included from src/pc1/../sealing/../util/thread_pool_t.hpp:13:0,
                 from src/pc1/../sealing/topology_t.hpp:10,
                 from src/pc1/pc1.hpp:6,
                 from src/pc1/pc1.cpp:31:
/usr/include/c++/7/atomic:737:7: note: declared here
       atomic(const atomic&) = delete;
       ^~~~~~
/usr/include/c++/7/atomic:741:17: note:   after user-defined conversion: constexpr std::atomic<long unsigned int>::atomic(std::atomic<long unsigned int>::__integral_type)
       constexpr atomic(__integral_type __i) noexcept : __base_type(__i) { }
                 ^~~~~~
+ ar rvs obj/libsupraseal.a obj/pc1.o obj/pc2.o obj/pc2_link.o obj/c1.o obj/supra_seal.o obj/column_reader.o obj/sector_parameters.o obj/sha_ext_mbx2.o
ar: creating obj/libsupraseal.a
ar: obj/pc1.o: No such file or directory

**Comment by dot-asm** [comment]
This looks very much like an out-of-date compiler. What's your system? Either way, there are two possible outcomes. We either try to make it work on older systems, or simply declare that it's supported on Ubuntu >=22 (or maybe >=20) or equivalent. Provided that it's a rather specialized application, it doesn't make much sense to invest an effort in the former...  It might be sensible to enforce a minimum C++ standard compliance level, but it would still be "if there still is a problem, point to the reference platform in the README.md."

**Comment by dot-asm** [comment]
> It might be sensible to enforce a minimum C++ standard compliance level

For reference, as following:
```diff --git a/build.sh b/build.sh
index 171b518..da083db 100755
--- a/build.sh
+++ b/build.sh
@@ -16,10 +16,12 @@ NVCC=${NVCC:-nvcc}
 CUDA=$(dirname $(dirname $(which $NVCC)))
 SPDK="deps/spdk-v22.09"
 CUDA_ARCH="-arch=sm_80 -gencode arch=compute_70,code=sm_70"
+CXXSTD=`$CXX -dM -E -x c++ /dev/null | \
+        awk '{ if($2=="__cplusplus" && $3<"2011") print "-std=c++11"; }'`
 
 INCLUDE="-I$SPDK/include -I$SPDK/isa-l/.. -I$SPDK/dpdk/build/include"
 CFLAGS="$SECTOR_SIZE -O2 -g $INCLUDE -D__ADX__"
-CPPFLAGS="$CFLAGS \
+CPPFLAGS="$CFLAGS $CXXSTD \
           -fno-omit-frame-pointer -Wall -Wextra -Wno-unused-parameter \
           -Wno-missing-field-initializers -fno-strict-aliasing \
           -march=native -Wformat -Wformat-security \

```

**Comment by dot-asm** [comment]
It's likely to be fixed by #6. Or at least there is hardly anything else we can offer.
